compiler/src/dmd/expressionsem.d

```diff
@@ -12319,6 +12319,29 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
         Type t1 = exp.e1.type;
         Type t2 = exp.e2.type;
+
+        // https://issues.dlang.org/show_bug.cgi?id=23767
+        // `cast(void*) 0` should be treated as `null` so the ternary expression
+        // gets the pointer type of the other branch
+        if (sc.flags & SCOPE.Cfile)
```

Comment:
Is there a way to change the way the types are merged rather than rewriting them? Not sure if you have the scope available but you could always just have a different function for finding the common types of C.

The C behaviour (in a sense) was almost what D used to do before I fixed it a while back.

Reply:
`void*` + `T*` merges to `void*` in general, both in C and D, only in C there is special treatment when the `void*` is the NULL constant

Reply:
Previously it went to T* until I fixed it (I think ldc 1.30 still has this bug)

Could potentially use TypeNull

Reply:
I am rewriting `(void*)0` to use `TypeNull`